### ISSUE TITLE
Enhancement: [no-explicit-any] Add option to allow `any` in satisfies operator

### ISSUE THREAD

**Issue by adamalfredsson** [typescript-eslint/typescript-eslint#11522]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-explicit-any/

## Current Behavior

The `@typescript-eslint/no-explicit-any` rule currently flags all uses of `any`, including when used within the `satisfies` operator. This creates friction when using `satisfies` for type checking where the specific types are intentionally flexible or complex.

## Desired Behavior

Add a new option to the `no-explicit-any` rule to allow `any` when used specifically within `satisfies` expressions.

## Use Case

The `satisfies` operator is often used to ensure a value conforms to a general shape without requiring exact type parameters. For example:

```typescript
type Routes = '/home' | '/about' | '/contact';

const routeConfig = {
  '/home': { component: HomeComponent, requiresAuth: false },
  '/about': { component: AboutComponent, requiresAuth: false },
  '/contact': { component: ContactComponent, requiresAuth: true },
} satisfies Record<Routes, any>;
```

In these cases:
- The actual types are correctly inferred from the values
- The `satisfies` clause only validates the general shape
- Using `unknown` doesn't work due to type constraints
- Specifying exact types would be redundant and harder to maintain

## Proposed Solution

Add a new option to the rule configuration:

```typescript
type Options = {
  ignoreRestArgs?: boolean;
  fixToUnknown?: boolean;
  allowInSatisfies?: boolean; // New option
}
```

Usage:
```json
{
  "rules": {
    "@typescript-eslint/no-explicit-any": ["error", {
      "allowInSatisfies": true
    }]
  }
}
```

### Fail

```TypeScript
type Routes = '/home' | '/about' | '/contact';

const routeConfig = {
  '/home': { component: HomeComponent, requiresAuth: false },
  '/about': { component: AboutComponent, requiresAuth: false },
  '/contact': { component: ContactComponent, requiresAuth: true },
} as Record<Routes, any>;
```

### Pass

```TypeScript
type Routes = '/home' | '/about' | '/contact';

const routeConfig = {
  '/home': { component: HomeComponent, requiresAuth: false },
  '/about': { component: AboutComponent, requiresAuth: false },
  '/contact': { component: ContactComponent, requiresAuth: true },
} satisfies Record<Routes, any>;
```

**Comment by bradzacher** [comment]
If there are no type constraints then you can use `unknown. If there are type constraints you can use `never`

Your examples don't actually show any type constraints though.

**Comment by adamkrupke** [comment]
You're right, it seems like `unknown` works well for covariant types demonstrated in these examples – sorry for not checking this. I had some success with using `never`, thanks! However, there's still one example where I neither `unknown` nor `any` is working:

```typescript
interface Container<T> {
  get: () => T;
  set: (value: T) => void;
}

const stringContainer: Container<string> = {
  get: () => "hello",
  set: (s: string) => console.log(s)
};

const numberContainer: Container<number> = {
  get: () => 10,
  set: (n: number) => console.log(n)
}

const containers = [
  stringContainer,
  numberContainer
] /* Check that this array only contains containers */ satisfies Container<any>[]
```

[Link to TS-playground](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgMIHtx1NAPAFQD5kBvAKGWQHMIwAuZACgEpkBeY-AbguQGdaDRgDc4AGwCuEBvlYdkw9MAAmPAL5kyCTHzD8wUUFQxYcUBibDYQeXYZBVibUrxr0mc4gCIAFhDFi6F4ANLwC7ox8DHZGnsjaIHzoYhAAdIFUkcxkajxaOnogEgC2AEbQltbQFphWZrhFZdBOLpRuQnEAjAAMoZThQiAMjeVQcQlJKenomSDZGvmJegl1NlB87MgA2mEGRpVmfcgjFbVVUGQAusgA9ABUaH4IANbIYD5weu-AG3BQUHAAJ7ITBiYErawbCFmDZ3G78T4-GDACAbA5rXBwECAwhbS6aIA)

Thanks!

**Comment by bradzacher** [comment]
In that rare case - you can use an `any` and a disable comment with it would be fine to explain the edge case.

I do not believe there is a common enough usecase to provide the necessary motivation here to add an option.

**Comment by adamkrupke** [comment]
Fair enough, thanks!
